docs/data/data-grid/column-menu/column-menu.md

```diff
@@ -0,0 +1,145 @@
+  />
+);
+```
+
+### Override default items and add new items:
+
+Using `componentsProps.columnMenu.columnMenuItems`, you can:
+
+1. Override Existing Items
+2. Add new Items
+
+```tsx
+const columnMenuItems = {
+  filter: <MyCustomFilter />, // override existing item
+  closeMenu: <MenuCloseComponent />, // add new item
+};
+
+// add new item in visible items and append it to the last of list
+const getVisibleColumnMenuItems = (defaultItems) => [...defaultItems, 'closeMenu'];
+
+<DataGrid
+  {...data}
+  componentProps={{ columnMenu: { getVisibleColumnMenuItems, columnMenuItems } }}
+/>;
+```
+
+If you're using TypeScript, for new items that you are adding, you'll need to specify new items you are registering, using [module augmentation](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation).
+
+```tsx
+declare module '@mui/x-data-grid' {
+  interface GridColumnMenuKeysOverrides {
+    closeMenu: true;
+  }
+}
+```
+
+### Default column menu items
+
+Default `keys` for **DataGrid** are `filter`, `sorting`, `hideColumn`, `divider` and `manageColumns`, wheras **DataGridPro** adds `pinning` and **DataGridPremium** adds `grouping` and `aggregation` on top of them.
+
+Here's a demo overriding some existing items, adding some new items and displaying different items for a column.
+
+{{"demo": "FilterColumnMenuGrid.js", "bg": "inline"}}
+
+## Column menu with Pro/Premium options [<span class="plan-pro"></span>](/x/introduction/licensing/#pro-plan)[<span class="plan-premium"></span>](/x/introduction/licensing/#premium-plan)
+
+You can access commercial features like column pinning etc from the column menu when using `DataGridPro` or `DataGridPremium`.
+
+{{"demo": "ColumnMenuGridPro.js", "bg": "inline"}}
+
+## Simple column menu
+
+You can replace the default column menu with a simplified one using `GridColumnMenuSimple` component.
+
+{{"demo": "SimpleColumnMenuGrid.js", "bg": "inline"}}
+
+For _Pro_ and _Premium_ packages you have specific `slot` for each of the column menu item using which you can import simple variant or pass on your own component, or you can also use [column menu customizing](#customise-column-menu-items) to use your custom component based on item `key`.
+
+Here's a list of column menu items and their overridable slots for Commercial features.
+
+| **Item**                | **Package** | **Key**       | **Slot**                     | **Default**                     |
+| ----------------------- | ----------- | ------------- | ---------------------------- | ------------------------------- |
+| Pinning                 | Pro         | 'pinning'     | 'ColumnMenuPinningItem'      | GridColumnPinningMenuItems      |
+| Groupable (not grouped) | Premium     | 'grouping'    | 'ColumnMenuRowGroupableItem' | GridRowGroupableColumnMenuItems |
+| Grouping (grouped)      | Premium     | 'grouping'    | 'ColumnMenuRowGroupingItem'  | GridRowGroupingColumnMenuItems  |
```

Comment:
Since the `key` is the same for both slots, how would this work with `columnMenuItems`?

Reply:
That's actually a good question, actually, at one point in time either `grouping` or `groupable` should be rendered, i.e for group column and the column that is grouped it should show `grouping`  component, and vice versa.

So there could be two possible solutions to this:
1. Let users have both keys but then they have to update the items returned by `getVisibleColumnMenuItems` on every group/ungroup. It seems fine but that might be more hassling for the user in default scenario and may also incur more re-renders due to `useCallback` being fired on every group/ungroup
2. I opted for this one to use one key `grouping` to handle both, it should work fine with our versions as we internally use `isGroupingColumn` to render the correct `slot` component, to override, the users can pass on respective `slot` components and leave the rendering part to us, if the user wants to configure this with `columnMenuItems` though, they can use the same utility `isGroupingColumn` like:

```tsx
import { isGroupingColumn } from '@mui/x-data-grid-premium' 
const CustomGroupComponent = ({ column }) => {
  if (isGroupingColumn(column.field) {
    return <CustomGroupingItem />
  }
  return <CustomGroupableItem />
}
```

And then we can use this alright to override the default item with the custom one using:
```tsx
const columnMenuItems = {
  grouping: <CustomGroupComponent />, // override 
}
```

I'm open to disagreements on this though :)

Reply:
Option 2 sounds good to me!
The reason I asked about `columnMenuItems` is that our default components do not stick to this pattern - instead of a single component there are 2 separate ones that are swapped during preprocessing:
 https://github.com/MBilalShafi/mui-x/blob/b4d4270640f439ffa107ecfbef6a6020e64a2001/packages/grid/x-data-grid-premium/src/hooks/features/rowGrouping/useGridRowGrouping.tsx#L166-L173
 
 > they can use the same utility isGroupingColumn
 
 What if we pass the result of `isGroupingColumn(column.field)` to grouping component?
 ```tsx
 const CustomGroupComponent = ({ isGroupingColumn }) => {
  if (isGroupingColumn) {
    return <CustomGroupingItem />
  }
  return <CustomGroupableItem />
}

 ```
 
 It's not easy to discover `isGroupingColumn`, passing prop could make things easier

Reply:
Agreed, usability wise that's a nice thing and will not anymore force users to import `isGroupingColumn` function, but design wise it requires some refactoring as right now all the `columnMenuItems` are being `React.cloneElement`'d from DataGrid package and we need to figure out a way to effectively be able to pass custom props for each `key`, especially those being injected from Pro and Premium packages

Do you think it makes sense to add a demo explaining the need of `isGroupingColumn` function now and take passing it as a boolean as an enhancement, which could be added in a follow-up PR?